test/utils/FunctionCallGenerator.sol

```diff
@@ -0,0 +1,177 @@
+
+    EnumerableSetLib.Bytes32Set internal _revokedKeyHashes;
+
+    constructor(address tokenA, address tokenB) {
+        _tokenA = tokenA;
+        _tokenB = tokenB;
+    }
+
+    function _randomTestKey() internal returns (TestKey memory newKey) {
+        newKey = TestKeyManager.withSeed(KeyType.Secp256k1, vm.randomUint());
+    }
+
+    function _registerCall(TestKey memory newKey) internal virtual returns (HandlerCall memory) {
+        return CallUtils.initHandlerDefault().withCall(CallUtils.encodeRegisterCall(newKey)).withCallback(
+            abi.encodeWithSelector(IHandlerGhostCallbacks.ghost_RegisterCallback.selector, newKey.toKey())
+        );
+    }
+
+    function _revokeCall(bytes32 keyHash) internal virtual returns (HandlerCall memory) {
+        return CallUtils.initHandlerDefault().withCall(CallUtils.encodeRevokeCall(keyHash)).withCallback(
+            abi.encodeWithSelector(IHandlerGhostCallbacks.ghost_RevokeCallback.selector, keyHash)
+        );
+    }
+
+    function _updateCall(bytes32 keyHash, Settings settings) internal virtual returns (HandlerCall memory) {
+        return CallUtils.initHandlerDefault().withCall(CallUtils.encodeUpdateCall(keyHash, settings)).withCallback(
+            abi.encodeWithSelector(IHandlerGhostCallbacks.ghost_UpdateCallback.selector, keyHash)
+        );
+    }
+
+    function _tokenTransferCall(address token, address to, uint256 amount)
+        internal
+        virtual
+        returns (HandlerCall memory)
+    {
+        return CallUtils.initHandlerDefault().withCall(
+            CallUtils.initDefault().withTo(token).withData(
+                abi.encodeWithSelector(ERC20.transfer.selector, to, amount)
+            )
+        );
+    }
+
+    // Ghost keys are persisted after the callbacks are triggered
+    function _getRandomGhostKeyHash(uint256 seed) internal view returns (bytes32) {
+        if (_ghostKeyHashes.values().length == 0) {
+            return bytes32(0);
+        }
+        return _ghostKeyHashes.values()[seed % _ghostKeyHashes.values().length];
+    }
+
+    function _getSettingsForKeyHash(bytes32 keyHash) internal virtual returns (Settings) {}
+
+    /**
+     * @notice Generate a random function call with equal weighting between function types
+     * @param randomSeed Random seed for generation
+     * @param depth Current recursion depth
+     * @return A call object for the generated function
+     */
+    function generateRandomFunctionHandlerCall(uint256 randomSeed, uint256 depth) public returns (HandlerCall memory) {
+        vm.assume(randomSeed < type(uint256).max);
+        // Select function type with equal weighting
+        uint256 functionType = _bound(randomSeed, 0, FUNCTION_COUNT - 1);
+
+        // REGISTER
+        if (functionType == FUNCTION_REGISTER || _ghostKeyHashes.values().length == 0) {
+            TestKey memory newKey = _randomTestKey();
+            if (newKey.toKeyHash() != bytes32(0)) {
+                return _registerCall(newKey);
```

Comment:
technically you could register the 0 address key, I wonder if we should test that edge case and make sure that registering the 0 address key with restricted settings, doesnt actually do anything

Reply:
yes I think this is better as an explicit test case